mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -3276,6 +3303,11 @@ Operation *LLVMDialect::materializeConstant(OpBuilder &builder, Attribute value,
   if (auto symbol = dyn_cast<FlatSymbolRefAttr>(value))
     if (isa<LLVM::LLVMPointerType>(type))
       return builder.create<LLVM::AddressOfOp>(loc, type, symbol);
+  // If this was folded from an llvm.mlir.zero operation, it should be
+  // materialized as such. However, prefer to use llvm.mlir.constant for
+  // compatible types.
```

Comment:
Since we fold integers / floats to integer and float attributes we probably don't need this special handling here and we can just do something like this:

```
if (isa<LLVM::ZeroAttr>(value))
  return builder.create<LLVM::ZeroOp>(loc, type);
if (isa<LLVM::UndefAttr>(value))
  return builder.create<LLVM::UndefOp>(loc, type);
if (isa<LLVM::PoisonAttr>(value))
  return builder.create<LLVM::PoisonOp>(loc, type);
```